glyphs-reader/src/propagate_anchors.rs

```diff
@@ -0,0 +1,883 @@
+//! Propagating anchors from components to their composites
+//!
+//! Glyphs.app has a nice feature where anchors defined in the components
+//! of composite glyphs are copied into the composites themselves. This feature
+//! is not very extensively documented, and the code here is based off the
+//! Objective-C implementation, which was shared with us privately.
+
+use std::collections::{BTreeMap, HashMap, VecDeque};
+
+use indexmap::IndexMap;
+use kurbo::{Affine, Vec2};
+use smol_str::{format_smolstr, SmolStr};
+
+use crate::{
+    font::Anchor,
+    glyphdata::{Category, Subcategory},
+    Component, Font, Glyph, Layer, Shape,
+};
+
+impl Font {
+    /// Copy anchors from component glyphs into their including composites
+    pub fn propagate_all_anchors(&mut self) {
+        propagate_all_anchors_impl(&mut self.glyphs);
+    }
+}
+
+// the actual implementation: it's easier to test a free fn
+fn propagate_all_anchors_impl(glyphs: &mut BTreeMap<SmolStr, Glyph>) {
+    // the reference implementation does this recursively, but we opt to
+    // implement it by pre-sorting the work to ensure we always process components
+    // first.
+    let todo = depth_sorted_composite_glyphs(glyphs);
+    let mut num_base_glyphs = HashMap::new();
+    // NOTE: there's an important detail here, which is that we need to call the
+    // 'anchors_traversing_components' function on each glyph, and save the returned
+    // anchors, but we only *set* those anchors on glyphs that have components.
+    // to make this work, we write the anchors to a separate data structure, and
+    // then only update the actual glyphs after we've done all the work.
+    let mut all_anchors = HashMap::new();
+    for name in todo {
+        // temporarily remove the glyph so we can mess with it without borrowing all glyphs
+        let glyph = glyphs.get(&name).unwrap();
+        // we work with indices here to also get around borrowck, because
+        // if we borrow the layers in the loop we can't set the anchors in the loop body
+        for layer in &glyph.layers {
+            let anchors = anchors_traversing_components(
```

Comment:
Optional, might just be me, I'd have grasped something like `reachable_anchors` faster

Reply:
This name was chosen to match the source impl, happy to change...